dev/src/ComponentManager.php

```diff
@@ -0,0 +1,192 @@
+
+    /**
+     * @param string $rootPath The path to the repository root directory.
+     * @param string $manifestPath [optional] The path to the manifest.json
+     *        file. May be left blank to use default location.
+     */
+    public function __construct($rootPath, $manifestPath = null)
+    {
+        $this->rootPath = $rootPath;
+        $this->manifestPath = $manifestPath
+            ?: $this->rootPath . self::DEFAULT_MANIFEST_PATH;
+    }
+
+    /**
+     * Return a list of component IDs.
+     *
+     * @return string[]
+     */
+    public function components()
+    {
+        $components = $this->components ?: $this->loadComponents();
+
+        return array_keys($components);
+    }
+
+    /**
+     * Get all component's composer.json extra data.
+     *
+     * @param string $componentId [optional] If set, only this component's data
+     *        will be returned.
+     * @return array[]
+     */
+    public function componentsExtra($componentId = null)
+    {
+        $components = $this->components ?: $this->loadComponents();
+
+        array_walk($components, function (&$component) {
+            $name = $component['composer']['name'];
+            $component = $component['composer']['extra']['component'];
+            $component['displayName'] = $name;
+        });
+
+        return $componentId
+            ? [$componentId => $components[$componentId]]
+            : $components;
+    }
+
+    /**
+     * Get all component's manifest.json data.
+     *
+     * @param string $componentId [optional] If set, only this component's data
+     *        will be returned.
+     * @return array[]
+     */
+    public function componentsManifest($componentId = null)
+    {
+        $manifest = $this->manifest ?: $this->loadManifest;
+        $modules = $manifest['modules'];
+
+        if ($componentId) {
+            $modules = array_filter($modules, function ($module) use ($componentId) {
+                return $module['id'] === $componentId;
+            });
+        }
+
+        return $modules;
+    }
+
+    /**
+     * Get the latest version of each component.
+     *
+     * @param string $componentId [optional] If set, only this component's data
+     *        will be returned.
+     * @return string[]
+     */
+    public function componentsVersion($componentId = null)
+    {
+        $components = $this->components ?: $this->loadComponents();
+
+        array_walk($components, function (&$component) {
+            $component = $component['version'];
+        });
+
+        return $componentId
+            ? [$componentId => $components[$componentId]]
+            : $components;
+    }
+
+    private function loadComponents()
+    {
+        $manifest = $this->manifest ?: $this->loadManifest();
+
+        $modules = $manifest['modules'];
+        $components = [];
+        foreach ($modules as $module) {
+            $components[$module['id']] = [
+                'version' => array_shift($module['versions']),
```

Comment:
We have a [unit test[(https://github.com/googleapis/google-cloud-php/blob/master/tests/Unit/JsonFileTest.php#L61-L68) which should verify the validity of the manifest file. It's also been quite some time since anything in the manifest was manually modified (in normal circumstances). Given that, I didn't find it necessary to validate the data here.